lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,385 @@
+    auto &rangeValsHandle{GetRangeValHandle<T>()};
+    auto selectCaseRangeValsIter{std::find_if(rangeValsHandle.begin(),
+        rangeValsHandle.end(), [&val, this](const auto &mem) {
+          return this->CheckOverlappingValue(val, mem.fromVal, mem.toVal);
+        })};
+    if (selectCaseRangeValsIter != rangeValsHandle.end()) {
+      ReportOverlapCaseValError(source, selectCaseRangeValsIter->source,
+          GetValueAsString(val).c_str());
+    }
+    auto selectCaseValsInsertStatus{GetValHandle<T>().emplace(val, source)};
+    if (!selectCaseValsInsertStatus.second) {
+      ReportOverlapCaseValError(source,
+          selectCaseValsInsertStatus.first->second,
+          GetValueAsString(val).c_str());
+    }
+  }
+  void Insert(bool logicalVal, parser::CharBlock source) {
+    auto selectCaseLogicalValsInsertStatus{
+        selectCaseLogicalVals_.emplace(logicalVal, source)};
+    if (!selectCaseLogicalValsInsertStatus.second) {
+      ReportOverlapCaseValError(source,
+          selectCaseLogicalValsInsertStatus.first->second,
+          GetValueAsString(logicalVal).c_str());
+    }
+  }
+  template<typename T>
+  void Insert(std::optional<T> lowerVal, std::optional<T> upperVal,
+      parser::CharBlock source) {
+    SelectCaseRangeType<T> rangeValue(lowerVal, upperVal, source);
+    auto &rangeValsHandle{GetRangeValHandle<T>()};
+    auto &valsHandle{GetValHandle<T>()};
+    auto selectCaseValsIter{std::find_if(valsHandle.begin(), valsHandle.end(),
+        [&rangeValue, this](const auto &mem) {
+          return this->CheckOverlappingValue(
+              mem.first, rangeValue.fromVal, rangeValue.toVal);
+        })};
+    if (selectCaseValsIter != valsHandle.end()) {
+      ReportOverlapCaseValError(source, selectCaseValsIter->second,
+          GetValuePairAsString(lowerVal, upperVal).c_str());
+    }
+    auto selectCaseRangeValsIter{std::find_if(rangeValsHandle.begin(),
+        rangeValsHandle.end(), [&rangeValue, this](const auto &mem) {
+          return ((!mem.fromVal && !rangeValue.fromVal) ||
+              (!mem.toVal && !rangeValue.toVal) ||
+              this->CheckOverlappingRangeValue(mem.fromVal, mem.toVal,
+                  rangeValue.fromVal, rangeValue.toVal));
+        })};
+    if (selectCaseRangeValsIter == rangeValsHandle.end()) {
+      rangeValsHandle.push_back(rangeValue);
+    } else {
+      ReportOverlapCaseValError(source, selectCaseRangeValsIter->source,
+          GetValuePairAsString(lowerVal, upperVal).c_str());
+    }
+  }
+  template<typename T> void InsertString(const T &x, parser::CharBlock source) {
+    if (kind_ == 1) {
+      if (auto val{GetStringValue(x)}) {
+        Insert(val.value(), source);
+      }
+    } else if (kind_ == 2) {
+      if (auto val{GetU16StringValue(x)}) {
+        Insert(val.value(), source);
+      }
+    } else if (kind_ == 4) {
+      if (auto val{GetU32StringValue(x)}) {
+        Insert(val.value(), source);
+      }
+    }
```

Comment:
assert/check here?

Reply:
I will add a check for default else case.

Reply:
I have added the default else case with proper error message.